SleepEngine/Engine/Object.h

```diff
@@ -4,55 +4,51 @@
 
 BeginNamespaceSleep
 
-class Object : public Updator
+class Object
 {
 public:
-    Object() noexcept(std::is_nothrow_constructible_v <Transform2D>);
-
-    Color getColor() const { return m_color; }
-    Color& modifyColor() { return m_color; }
-    void setColor(Color const color) { m_color = color; }
-
-    glm::vec2 getScale() const { return m_transform.scale; }
-    void setScale(glm::vec2 const scale) { m_transform.scale = scale; }
-
-    glm::vec2 getPosition() const { return m_transform.position; }
-    void setPosition(glm::vec2 const position) { m_transform.position = position; }
-
-    glm::vec2 getSize() const { return m_transform.size; }
-    void setSize(glm::vec2 const size) { m_transform.size = size; }
-
-    float getX() const { return m_transform.position.x; }
-    void setX(float x) { m_transform.position.x = x; }
-
-    float getY() const { return m_transform.position.y; }
-    void setY(float y) { m_transform.position.y = y; }
-
-    Texture* getTexture() const { return m_texture; }
-    void setTexture(Texture* const texture) { m_texture = texture; }
-
-    Layer getLayer() const { return m_transform.layer; }
-    void setLayer(Layer layer);
-
-    glm::vec2 getTopLeftUV() const { return m_topLeftUV; }
-    void setTopLeftUV(glm::vec2 const uv) { m_topLeftUV = uv; }
-
-    glm::vec2 getDownRightUV() const { return m_downRightUV; }
-    void setDownRightUV(glm::vec2 const uv) { m_downRightUV = uv; }
-
-    void setUV(float topLeftX, float topLeftY, float downRightX, float downRightY);
-
-    float getRotation() const { return m_transform.rotation; }
-    void setRotation(float const rotation) { m_transform.rotation = rotation; }
-
-    void render() override;
-
-private:
-    Transform2D m_transform;
-    Color m_color;
-    glm::vec2 m_topLeftUV;
-    glm::vec2 m_downRightUV;
-    NotOwnedPtr <Texture> m_texture;
+    using ComponentsContainer = std::vector <std::unique_ptr <Component>>;
+
+    virtual void update(float dt);
+
+    void addComponent(ComponentsContainer::value_type&& component);
+
+    template <class TComponent>
+    void addComponent()
+    {
+        static_assert(std::is_default_constructible_v <TComponent>, "TComponent is not default constructible. Please, "
+            "create an instance of this class and use non-templated 'addConponent' method.");
+        addComponent(std::make_unique <TComponent>());
+    }
+    
+    template <class TComponent>
+    TComponent const* getComponent() const
+    {
+        auto const id = Component::getComponentTypeID <TComponent>();
+
+        auto const findByID = [id](auto const& component)
+        {
+            return component->getComponentTypeID() == id;
+        };
```

Comment:
Technically it doesn't find anything, technically. Its more like checking, but don't sewat it.

Reply:
Why? I really need to find an object with such id in my vector of components.

Reply:
Unfortunately, we must store this ID in every component to be able to find an exact component class by its id even when it is being stored as a Component interface.
There is no overhead because anyway we had to store some other ID (e.g. enum with all components).
Is this what you meant?

Reply:
nope i meant that function doest not __find__ anything, it's comparing 2 components

Reply:
Yeah, I gave it this name because it suits the find_if call. It looks like:
find_if (All(vector), findByID)